File: src/modules/trading/format.js
/**
 * @file Number formatters for trading module display.
 * Manual VND formatter avoids locale-dependent toLocaleString in CF Workers.
 */

/**
 * Format number as Vietnamese Dong — dot thousands separator, no decimals.
 * @param {number} n
 * @returns {string} e.g. "15.000.000 VND"
 */
export function formatVND(n) {
  const rounded = Math.round(n);
  const abs = Math.abs(rounded).toString();
  // insert dots every 3 digits from right
  let result = "";
  for (let i = 0; i < abs.length; i++) {
    if (i > 0 && (abs.length - i) % 3 === 0) result += ".";
    result += abs[i];
  }
  return `${rounded < 0 ? "-" : ""}${result} VND`;
}

/**
 * Format number as USD — 2 decimals, comma thousands.
 * @param {number} n
 * @returns {string} e.g. "$1,234.56"
 */
export function formatUSD(n) {
  const fixed = Math.abs(n).toFixed(2);
  const [intPart, decPart] = fixed.split(".");
  let result = "";
  for (let i = 0; i < intPart.length; i++) {
    if (i > 0 && (intPart.length - i) % 3 === 0) result += ",";
    result += intPart[i];
  }
  return `${n < 0 ? "-" : ""}$${result}.${decPart}`;
}

/**
 * Format crypto quantity — up to 8 decimals, trailing zeros stripped.
 * @param {number} n
 * @returns {string} e.g. "0.00125"
 */
export function formatCrypto(n) {
  return Number.parseFloat(n.toFixed(8)).toString();
}

/**
 * Format stock quantity — integer only.
 * @param {number} n
 * @returns {string}
 */
export function formatStock(n) {
  return Math.floor(n).toString();
}

/**
 * Format amount based on currency type.
 * @param {number} n
 * @param {string} currency — "VND" or "USD"
 * @returns {string}
 */
export function formatCurrency(n, currency) {
  if (currency === "VND") return formatVND(n);
  if (currency === "USD") return formatUSD(n);
  return `${n} ${currency}`;
}

/**
 * Format P&L line with absolute and percentage.
 * @param {number} currentValue
 * @param {number} invested
 * @returns {string}
 */
export function formatPnL(currentValue, invested) {
  const diff = currentValue - invested;
  const pct = invested > 0 ? ((diff / invested) * 100).toFixed(2) : "0.00";
  const sign = diff >= 0 ? "+" : "";
  return `${sign}${formatVND(diff)} (${sign}${pct}%)`;
}

File: internal/modules/stock/format.go
// Package stock is a paper-stock module for VN stocks. It keeps a per-user
// live portfolio in KV and prices positions at the current market price.
package stock

import (
	"math"
	"strconv"
	"strings"
)

// FormatVND renders an integer-rounded amount with dot-thousands separators
// and a "VND" suffix (e.g. 15000000 -> "15.000.000 VND"). Manual to avoid
// locale-dependent formatting.
func FormatVND(n float64) string {
	return formatVNDNumber(n) + " VND"
}

// formatVNDNumber renders a VND amount without its currency suffix. Portfolio
// tables use this because their title declares the currency once.
func formatVNDNumber(n float64) string {
	rounded := int64(math.Round(n))
	abs := strconv.FormatInt(absInt64(rounded), 10)
	var sb strings.Builder
	if rounded < 0 {
		sb.WriteByte('-')
	}
	for i := 0; i < len(abs); i++ {
		if i > 0 && (len(abs)-i)%3 == 0 {
			sb.WriteByte('.')
		}
		sb.WriteByte(abs[i])
	}
	return sb.String()
}

// FormatStock renders an integer share quantity (always whole shares).
func FormatStock(n float64) string {
	return strconv.FormatInt(int64(math.Floor(n)), 10)
}

// formatShareQuantity renders an exact whole-share quantity with Vietnamese
// dot-thousands separators, without converting the int64 value through float64.
func formatShareQuantity(n int64) string {
	raw := strconv.FormatInt(n, 10)
	digitStart := 0
	if raw[0] == '-' {
		digitStart = 1
	}

	var sb strings.Builder
	sb.Grow(len(raw) + (len(raw)-digitStart-1)/3)
	if digitStart == 1 {
		sb.WriteByte('-')
	}
	for i := digitStart; i < len(raw); i++ {
		if i > digitStart && (len(raw)-i)%3 == 0 {
			sb.WriteByte('.')
		}
		sb.WriteByte(raw[i])
	}
	return sb.String()
}

// FormatPnL renders a signed VND delta + percentage line, e.g.
// "+1.234 VND (+12.34%)" or "-500.000 VND (-5.00%)". When invested is zero
// the percentage is reported as 0.00 to avoid division-by-zero.
func FormatPnL(currentValue, invested float64) string {
	return formatPnL(currentValue, invested, FormatVND)
}

func formatPortfolioPnL(currentValue, invested float64) string {
	return formatPnL(currentValue, invested, formatVNDNumber)
}

func formatPnL(currentValue, invested float64, formatAmount func(float64) string) string {
	diff := currentValue - invested
	pct := 0.0
	if invested > 0 {
		pct = (diff / invested) * 100
	}
	sign := ""
	if diff >= 0 {
		sign = "+"
	}
	return sign + formatAmount(diff) + " (" + sign + strconv.FormatFloat(pct, 'f', 2, 64) + "%)"
}

func absInt64(n int64) int64 {
	if n < 0 {
		return -n
	}
	return n
}
